src/CloudFoundry.WinDEA/StagingInstance.cs

```diff
@@ -265,34 +267,226 @@ public void GetBuildpack(StagingStartMessageRequest message, string gitPath, str
                     throw new Exception("Buildpack does not support this application.");
                 }
             }
+            else if (message.Properties.BuildpackKey != null)
+            {
+                Logger.Info("Staging task {0}: Using admin buildpack {1}", this.Properties.TaskId, message.Properties.BuildpackKey);
+
+                TryCleanupUnrackedAdminBuildpacks(message.AdminBuildpacks, adminBuildpackDir);
+                var ab = message.AdminBuildpacks.First(i => i.Key == message.Properties.BuildpackKey);
+
+                var buildpackMutex = new Mutex(false, "admin-buildpack-" + ab.Key);
+                try
+                {
+                    buildpackMutex.WaitOne();
+
+                    DownloadAdminBuildpack(ab, adminBuildpackDir);
+
+                    var buildpackDir = Path.Combine(adminBuildpackDir, ab.Key);
+
+                    DEAUtilities.DirectoryCopy(buildpackDir, Path.Combine(this.Workspace.TempDir, "buildpack"), true);
+                }
+                finally
+                {
+                    buildpackMutex.ReleaseMutex();
+                    buildpackMutex.Dispose();
+                }
+
+                this.Buildpack = new Buildpack(Path.Combine(this.Workspace.TempDir, "buildpack"), Path.Combine(this.Workspace.StagedDir, "app"), this.Workspace.Cache, this.Workspace.StagingLogPath);
+
+                bool detected = this.Buildpack.Detect(this.Container);
+                if (!detected)
+                {
+                    throw new Exception("Buildpack does not support this application.");
+                }
+            }
             else
             {
                 Logger.Info("Staging task {0}: Detecting buildpack", this.Properties.TaskId);
-                foreach (string dir in Directory.EnumerateDirectories(buildpacksDir))
+
+                List<string> systemBuildpacks = Directory.EnumerateDirectories(buildpacksDir).ToList();
+
+                if (message.AdminBuildpacks != null)
                 {
-                    DEAUtilities.DirectoryCopy(dir, Path.Combine(this.Workspace.TempDir, "buildpack"), true);
-                    Buildpack bp = new Buildpack(Path.Combine(this.Workspace.TempDir, "buildpack"), Path.Combine(this.Workspace.StagedDir, "app"), this.Workspace.Cache, this.Workspace.StagingLogPath);
-                    bool success = bp.Detect(this.Container);
-                    if (success)
+                    TryCleanupUnrackedAdminBuildpacks(message.AdminBuildpacks, adminBuildpackDir);
+
+                    foreach (var ab in message.AdminBuildpacks)
                     {
-                        this.Buildpack = bp;
-                        break;
+                        var dir = Path.Combine(adminBuildpackDir, ab.Key);
+
+                        var buildpackMutex = new Mutex(false, "admin-buildpack-" + ab.Key);
```

Comment:
can you extract this to a method (create mutex, wait and download)?
If not, at least create a subroutine that gets the mutex name, so it can be reused.